overlord/snapstate/snapmgr.go

```diff
@@ -564,7 +564,17 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	}
 
 	st.Unlock()
+	// XXX: this block is slightly ugly, find a pattern when we have more examples
 	err = m.backend.LinkSnap(newInfo)
+	if err != nil {
```

Comment:
I may be missing something, but the logic here doesn't look right. It's undoing in-place, logging an error, and then proceeding as if everything was fine. I haven't checked the tests, but they suspect they might not be good enough too if they're accepting this situation.

Let's please talk online about this.


I'd like to talk about it live when you have a moment.